### PR TITLE
Convert PrePrepareMsg to smart pointer

### PR DESCRIPTION
**Pull request by Efrat1**
Convert PrePrepareMsg raw pointer to smart in:
- ReplicaImp::onMessage<PrePrepareMsg>()
- startConsensusProcess().

Tested: Apollo, maestro, ASAN

### CODE REVIEW

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 1]
```diff
@@ -920,25 +919,27 @@ void ReplicaImp::startConsensusProcess(PrePrepareMsg *pp, bool isCreatedEarlier)
   SeqNumInfo &seqNumInfo = mainLog->get(primaryLastUsedSeqNum);
   {
     TimeRecorder scoped_timer1(*histograms_.addSelfMsgPrePrepare);
-    seqNumInfo.addSelfMsg(pp);
+    seqNumInfo.addSelfMsg(std::move(pp));
   }
+  // pp was moved, use ppMsg for next (not owning) operations
+  auto ppMsg = seqNumInfo.getPrePrepareMsg();
```
Comment (glevkovich) [inline bftengine/src/bftengine/ReplicaImp.cpp]: 1) nit:
Why do you need to call to another object to get a pointer which you already have? that looks a bit strange to me. how about:
```
 SeqNumInfo &seqNumInfo = mainLog->get(primaryLastUsedSeqNum);
auto ppMsg = pp.get();
  {
    TimeRecorder scoped_timer1(*histograms_.addSelfMsgPrePrepare);
    seqNumInfo.addSelfMsg(std::move(pp));
  }

if (ps_) { ...
```

2)
Do we have a task to convert seqNumInfo and possibly mainLog into a smart pointer?
getPrePrepareMsg returns a raw pointer - how much effort to add this into your task? I think too much?
If yes, lets open a task.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 2]
```diff
@@ -1082,32 +1083,29 @@ void ReplicaImp::onMessage<PrePrepareMsg>(PrePrepareMsgUPtr message) {
   SCOPED_MDC_SEQ_NUM(std::to_string(msgSeqNum));
   LOG_INFO(CNSUS, "Received PrePrepareMsg" << KVLOG(message->senderId(), msgSeqNum, message->size()));
 
-  auto *msg = message.release();
   if (!currentViewIsActive() && viewsManager->waitingForMsgs() && msgSeqNum > lastStableSeqNum) {
-    ConcordAssert(!msg->isNull());  // we should never send (and never accept) null PrePrepare message
+    ConcordAssert(!message->isNull());  // we should never send (and never accept) null PrePrepare message
 
-    if (viewsManager->addPotentiallyMissingPP(msg, lastStableSeqNum)) {
+    if (viewsManager->addPotentiallyMissingPP(message.release(), lastStableSeqNum)) {
```
Comment (glevkovich) [inline bftengine/src/bftengine/ReplicaImp.cpp]: 1) 
This is a leak. You can't release before you got the answer from addPotentiallyMissingPP.
If addPotentiallyMissingPP returns false you must keep ownership:
```
if (viewsManager->addPotentiallyMissingPP(message.get(), lastStableSeqNum)) {
  message.release(); 
   ...
```

2) Do we have a task to convert views manager (all functions, or at lease this one, depends on its size) to smart pointers?

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 3]
```diff
@@ -1117,11 +1115,13 @@ void ReplicaImp::onMessage<PrePrepareMsg>(PrePrepareMsgUPtr message) {
     // For MDC it doesn't matter which type of fast path
     SCOPED_MDC_PATH(CommitPathToMDCString(slowStarted ? CommitPath::SLOW : CommitPath::OPTIMISTIC_FAST));
     // All pre-prepare messages are added in seqNumInfo even if replica detects incorrect time
-    if (seqNumInfo.addMsg(msg, false, time_is_ok)) {
+    if (seqNumInfo.addMsg(std::move(message), false, time_is_ok)) {
```
Comment (glevkovich) [inline bftengine/src/bftengine/ReplicaImp.cpp:1121]:
```
auto ppMsg = message.get();
if (seqNumInfo.addMsg(std::move(message), false, time_is_ok)) {
```

and then use ppMsg and avoid calling  seqNumInfo.getPrePrepareMsg(); in line 1121 and 1139.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 4]
```diff
@@ -1544,7 +1546,7 @@ void ReplicaImp::onInternalMsg(InternalMessage &&msg) {
       ppm = nullptr;
       return;
     } else {
-      return startConsensusProcess(*ppm, true);
+      return startConsensusProcess(PrePrepareMsgUPtr(*ppm), true);
```
Comment (glevkovich) [inline bftengine/src/bftengine/ReplicaImp.cpp]: nit: use make_unique, it is the recommended way vs uptr ctor.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 5]
```diff
@@ -3191,23 +3193,24 @@ void ReplicaImp::onNewView(const std::vector<PrePrepareMsg *> &prePreparesForNew
...
+    ppSeqNumbers.push_back(pp->seqNumber());
 
     if (ps_) {
       ps_->setPrePrepareMsgInSeqNumWindow(pp->seqNumber(), pp);
       ps_->setSlowStartedInSeqNumWindow(pp->seqNumber(), true);
     }
 
     if (primaryIsMe)
-      seqNumInfo.addSelfMsg(pp);
+      seqNumInfo.addSelfMsg(std::make_unique<PrePrepareMsg>(pp));
```
Comment (glevkovich) [inline bftengine/src/bftengine/ReplicaImp.cpp:3212]: This looks not right.
It all starts from the fact that we pass into this function a vector with raw pointers. If the vector was with a smart pointer, why would you need to create a new smart poiunter here and not just move an existing one?

So this code and all the later parts which create unique_ptr looks artificial ("half work"? )
now, if you hve future tasks it's fine, although thye must be documented very well.
But I assume you dont.
Start from passing 
`const std::vector<PrePrepareMsgUptr> &prePreparesForNewView`

and if you dont want to do it here, open a new task. But this code should just move the pointer in prePreparesForNewView and not create a raw pointer.
same for all below code till function end.

File: bftengine/src/bftengine/SeqNumInfo.cpp [hunk 6]
```diff
@@ -83,39 +82,38 @@ void SeqNumInfo::resetAndFree() {
 }
 
 void SeqNumInfo::getAndReset(PrePrepareMsg*& outPrePrepare, PrepareFullMsg*& outCombinedValidSignatureMsg) {
-  outPrePrepare = prePrepareMsg;
-  prePrepareMsg = nullptr;
+  outPrePrepare = prePrepareMsg.release();
```
Comment (glevkovich) [inline bftengine/src/bftengine/SeqNumInfo.cpp]: Strange function definition style. That's how C is written, not modern C++.
I suggest to transform (consider if you like in another task):
`std::pair<PrePrepareMsg*, PrepareFullMsg*>  SeqNumInfo::getAndReset()`

caller can just call:
`auto [ ppMsg, ppFMsg ] = seqNumInfo::getAndReset();`

**Review by glevkovich — CHANGES_REQUESTED** [review]

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 1 above)
Comment (Efrat1) [reply bftengine/src/bftengine/ReplicaImp.cpp]: 1. Changed, thanks.
2. PrePrepareMsg in SeqNumInfo is already a unique_ptr, and SeqNumInfo is the owner of it. getPrePrepareMsg and getSelfPrePrepareMsg return raw pointer as they are used to read only.
I created a task for other raw pointers in SeqNumInfo: [BC-26948](https://jira.eng.vmware.com/browse/BC-26948)

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 2 above)
Comment (Efrat1) [reply bftengine/src/bftengine/ReplicaImp.cpp]: 1. Thanks, fixed.
2. Yes, [BC-26902](https://jira.eng.vmware.com/browse/BC-26902)

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 3 above)
Comment (Efrat1) [reply bftengine/src/bftengine/ReplicaImp.cpp:1121]: Done

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 4 above)
Comment (Efrat1) [reply bftengine/src/bftengine/ReplicaImp.cpp]: Done, thanks.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 5 above)
Comment (Efrat1) [reply bftengine/src/bftengine/ReplicaImp.cpp:3212]: I agree.
The vector of raw pointers (prePreparesForNewView) came from ViewManager which holds only raw pointers.
I did open a task for ViewManager ([BC-26902](https://jira.eng.vmware.com/browse/BC-26902)). However, I edited its description to be more explicit.

File: bftengine/src/bftengine/SeqNumInfo.cpp (hunk 6 above)
Comment (Efrat1) [reply bftengine/src/bftengine/SeqNumInfo.cpp]: Changed, thanks.

**Review by glevkovich — APPROVED** [review]
LGTM, thanks for the hard work!

**Pull request merged**